src/websocket_client.rs

```diff
@@ -0,0 +1,171 @@
+                                            self_clone.executor_handle.spawn(proxy);
+                                        }).map_err(|e| error!("upgrade error: {}", e));
+
+                                        self.executor_handle.spawn(fut);
+
+                                        return Box::new(futures::future::ok::<Response<Body>, hyper::Error>(res));
+                                    } else {
+                                        *response.status_mut() = StatusCode::PRECONDITION_REQUIRED;
+                                        return Box::new(futures::future::ok::<Response<Body>, hyper::Error>(response));
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+                *response.status_mut() = StatusCode::BAD_REQUEST;
+            } else if req.uri().path().starts_with("/jet/create") {
+                let uuid = Uuid::new_v4();
+                if let Ok(mut jet_associations) = self.jet_associations.try_lock() {
+                    jet_associations.insert(uuid, None);
+                    *response.body_mut() = Body::from(uuid.to_string())
+                }
+            } else {
+                *response.status_mut() = StatusCode::BAD_REQUEST;
+            },
+
+            _ => {
+                *response.status_mut() = StatusCode::BAD_REQUEST;
+            }
+        }
+
+        Box::new(futures::future::ok::<Response<Body>, hyper::Error>(response))
+    }
+}
+
+
+fn process_req(req: &Request<Body>) -> Response<Body> {
+    /*
+        Source: https://gist.github.com/bluetech/192c74b9c4ae541747718ac4f4e20a14
+        Author: Ran Benita<bluetech> (ran234@gmail.com)
+    */
+
+    fn convert_key(input: &[u8]) -> String {
+        const WS_GUID: &[u8] = b"258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+        let mut digest = sha1::Sha1::new();
+        digest.update(input);
+        digest.update(WS_GUID);
+        base64::encode(&digest.digest().bytes())
+    }
+    fn connection_has(value: &header::HeaderValue, needle: &str) -> bool {
+        if let Ok(v) = value.to_str() {
+            v.split(',').any(|s| s.trim().eq_ignore_ascii_case(needle))
+        } else {
+            false
+        }
+    }
+    let is_http_11 = req.version() == Version::HTTP_11;
+    let is_upgrade = req.headers()
+        .get(header::CONNECTION)
+        .map_or(false, |v| connection_has(v, "upgrade"));
+    let is_websocket_upgrade = req.headers()
+        .get(header::UPGRADE)
+        .and_then(|v| v.to_str().ok())
+        .map_or(false, |v| v.eq_ignore_ascii_case("websocket"));
+    let is_websocket_version_13 = req.headers()
+        .get(header::SEC_WEBSOCKET_VERSION)
+        .and_then(|v| v.to_str().ok())
+        .map_or(false, |v| v == "13");
```

Comment:
checking for websocket version 13 explicitly is probably too restrictive. maybe we should just check for the websocket version header presence instead, unless we have a meaningful check to perform with regards to the version number.

Reply:
This version check can be removed entirely if this is a cause for concern

Reply:
V13 was standardized in December 2011 and is supported from IE 10 and above and all other chromium base / Mozilla browser support it, the check could instead be replaced by a check for a version greater or equal to v13.

Reply:
I agree with this suggestion. Let's parse and check for version >= 13.